xds/internal/xdsclient/xdsresource/test/unmarshal_cds_test.go

```diff
@@ -0,0 +1,609 @@
+						MaximumRingSize: wrapperspb.UInt64(200),
+					},
+				},
+				LoadBalancingPolicy: &v3clusterpb.LoadBalancingPolicy{
+					Policies: []*v3clusterpb.LoadBalancingPolicy_Policy{
+						{
+							TypedExtensionConfig: &v3corepb.TypedExtensionConfig{
+								TypedConfig: testutils.MarshalAny(&v3ringhashpb.RingHash{
+									HashFunction:    v3ringhashpb.RingHash_XX_HASH,
+									MinimumRingSize: wrapperspb.UInt64(10),
+									MaximumRingSize: wrapperspb.UInt64(100),
+								}),
+							},
+						},
+					},
+				},
+			},
+			wantUpdate: xdsresource.ClusterUpdate{
+				ClusterName: clusterName, EDSServiceName: serviceName,
+			},
+			wantLBConfig: &internalserviceconfig.BalancerConfig{
+				Name: "ring_hash_experimental",
+				Config: &ringhash.LBConfig{
+					MinRingSize: 20,
+					MaxRingSize: 200,
+				},
+			},
+			customLBDisabled: true,
+		},
+		{
+			name: "load-balancing-policy-takes-precedence-over-lb-policy-and-enum",
+			cluster: &v3clusterpb.Cluster{
+				Name:                 clusterName,
+				ClusterDiscoveryType: &v3clusterpb.Cluster_Type{Type: v3clusterpb.Cluster_EDS},
+				EdsClusterConfig: &v3clusterpb.Cluster_EdsClusterConfig{
+					EdsConfig: &v3corepb.ConfigSource{
+						ConfigSourceSpecifier: &v3corepb.ConfigSource_Ads{
+							Ads: &v3corepb.AggregatedConfigSource{},
+						},
+					},
+					ServiceName: serviceName,
+				},
+				LbPolicy: v3clusterpb.Cluster_RING_HASH,
+				LbConfig: &v3clusterpb.Cluster_RingHashLbConfig_{
+					RingHashLbConfig: &v3clusterpb.Cluster_RingHashLbConfig{
+						MinimumRingSize: wrapperspb.UInt64(20),
+						MaximumRingSize: wrapperspb.UInt64(200),
+					},
+				},
+				LoadBalancingPolicy: &v3clusterpb.LoadBalancingPolicy{
+					Policies: []*v3clusterpb.LoadBalancingPolicy_Policy{
+						{
+							TypedExtensionConfig: &v3corepb.TypedExtensionConfig{
+								TypedConfig: testutils.MarshalAny(&v3ringhashpb.RingHash{
+									HashFunction:    v3ringhashpb.RingHash_XX_HASH,
+									MinimumRingSize: wrapperspb.UInt64(10),
+									MaximumRingSize: wrapperspb.UInt64(100),
+								}),
+							},
+						},
+					},
+				},
+			},
+			wantUpdate: xdsresource.ClusterUpdate{
+				ClusterName: clusterName, EDSServiceName: serviceName,
+			},
+			wantLBConfig: &internalserviceconfig.BalancerConfig{
+				Name: "ring_hash_experimental",
+				Config: &ringhash.LBConfig{
+					MinRingSize: 10,
+					MaxRingSize: 100,
+				},
+			},
+		},
+	}
+
+	oldAggregateAndDNSSupportEnv := envconfig.XDSAggregateAndDNS
+	envconfig.XDSAggregateAndDNS = true
+	defer func() { envconfig.XDSAggregateAndDNS = oldAggregateAndDNSSupportEnv }()
+	oldRingHashSupport := envconfig.XDSRingHash
+	envconfig.XDSRingHash = true
+	defer func() { envconfig.XDSRingHash = oldRingHashSupport }()
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			if test.customLBDisabled {
+				envconfig.XDSCustomLBPolicy = false
+				defer func() {
+					envconfig.XDSCustomLBPolicy = true
```

Comment:
Should this be `envconfig.XDSCustomLBPolicy = oldCustomLBSupport` instead?

Reply:
That would set it to the wrong value since I set the value at the beginning of the test :D. Let me know if you see a cleaner way of doing this.